### PR TITLE
Feature/experiment cache decorator

### PR DESCRIPTION
**Pull request by tomasjurasek**
### Motivation

Which issue does this fix? Fixes #`issue number`

If no issue exists, what is the fix or new feature? Were there any reasons to fix/implement things that are not obvious?

### Checklist

- [ ] Code follows coding conventions held in this repo
- [ ] Automated tests have been added
- [ ] Tests are passing
- [ ] Docs have been updated (if applicable)
- [ ] Temporary settings (e.g. variables used during development and testing) have been reverted to defaults

### How to test

If manual testing is required, what are the steps?

### CODE REVIEW

File: Kentico.Kontent.Delivery.Tests/Extensions/FakeServiceCollection.cs [hunk 1]
```diff
@@ -1,172 +0,0 @@
...
-using Microsoft.Extensions.DependencyInjection;
-using Microsoft.Extensions.Options;
-
-namespace Kentico.Kontent.Delivery.Tests.Extensions
-{
-    /// <summary>
-    /// Mock ServiceCollection that keeps inside everything that was registered.
-    /// Can be used to check whether expected instances were added to the ServiceCollection.
-    /// </summary>
-    internal class FakeServiceCollection : IServiceCollection
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery.Tests/Extensions/FakeServiceCollection.cs]: why did you decide to remove it?

File: Kentico.Kontent.Delivery/DeliveryClient.cs [hunk 2]
```diff
@@ -88,6 +88,8 @@ public async Task<JObject> GetItemJsonAsync(string codename, params string[] par
             return (await GetDeliverResponseAsync(endpointUrl)).Content;
         }
 
+
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery/DeliveryClient.cs]: please make sure you remove all unnecessary lines

File: Kentico.Kontent.Delivery/DeliveryClientFactory.cs [hunk 3]
```diff
@@ -42,12 +42,8 @@ public IDeliveryClient Get(string name)
                 throw new ArgumentNullException(nameof(name));
             }
 
-            if (!_cachedDeliveryClients.TryGetValue(name, out var client))
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery/DeliveryClientFactory.cs]: ❤

File: Kentico.Kontent.Delivery/DeliveryClientFactory.cs [hunk 4]
```diff
@@ -13,7 +13,7 @@ namespace Kentico.Kontent.Delivery
     /// <summary>
     /// A factory class for <see cref="IDeliveryClient"/>
     /// </summary>
-    public class DeliveryClientFactory  : IDeliveryClientFactory
+    public class DeliveryClientFactory : IDeliveryClientFactory
     {
         private ConcurrentDictionary<string, IDeliveryClient> _cachedDeliveryClients = new ConcurrentDictionary<string, IDeliveryClient>();
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery/DeliveryClientFactory.cs]: not used

File: Kentico.Kontent.Delivery/Configuration/DeliveryOptions.cs [hunk 5]
```diff
@@ -1,4 +1,6 @@
-﻿namespace Kentico.Kontent.Delivery
+﻿using Kentico.Kontent.Delivery.Configuration;
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery/Configuration/DeliveryOptions.cs]: doesn't seem to be necessary

File: Kentico.Kontent.Delivery.Caching/DeliveryClientCache.cs [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+using System.Threading.Tasks;
+
+namespace Kentico.Kontent.Delivery.Caching
+{
+    public class DeliveryClientCache : IDeliveryClient
+    {
+        private IDeliveryClient DeliveryClient { get; }
+        private IDeliveryCacheManager DeliveryCacheManager { get; }
+
+        public DeliveryClientCache(IDeliveryCacheManager cacheManager, IDeliveryClient deliveryClient)
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery.Caching/DeliveryClientCache.cs]: public members are missing comments

Comment (petrsvihlik) [reply Kentico.Kontent.Delivery.Caching/DeliveryClientCache.cs]: please check the warning tab, there are about 50 occurrences

File: Kentico.Kontent.Delivery/Builders/DeliveryClient/IDeliveryClientBuilder.cs [hunk 7]
```diff
@@ -35,6 +35,12 @@ public interface IOptionalClientSetup : IDeliveryClientBuild
         /// <param name="deliveryHttpClient">A custom <see cref="IDeliveryHttpClient"/> implementation</param>
         IOptionalClientSetup WithDeliveryHttpClient(IDeliveryHttpClient deliveryHttpClient);
 
+        /// <summary>
+        /// Use a custom cache manager
+        /// </summary>
+        /// <param name="deliveryCacheManager">A custom <see cref="IDeliveryCacheManager"/> implementation</param>
+        IOptionalClientSetup WithDeliveryCacheManager(IDeliveryCacheManager deliveryCacheManager);
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery/Builders/DeliveryClient/IDeliveryClientBuilder.cs]: not clear to me how this is supposed to be used...?

File: Kentico.Kontent.Delivery.Tests/Extensions/ServiceCollectionsExtensionsTests.cs [hunk 8]
```diff
@@ -63,92 +50,84 @@ public class ServiceCollectionsExtensionsTests
...
+        public static IEnumerable<object[]> DeliveryOptionsConfigurationParameters =>
+           new[]
+           {
+                new[] {"as_root"},
+                new[] {"under_default_key", "DeliveryOptions"},
+                new[] {"under_custom_key", "CustomNameForDeliveryOptions"},
+                new[] {"nested_under_default_key", "Options:DeliveryOptions"}
+           };
 
-        private readonly FakeServiceCollection _fakeServiceCollection;
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery.Tests/Extensions/ServiceCollectionsExtensionsTests.cs]: the same question as above - why did we have to remove this, and consequently the tests?

Comment (petrsvihlik) [reply Kentico.Kontent.Delivery.Tests/Extensions/ServiceCollectionsExtensionsTests.cs]: it seems that we're missing quite a lot of test scenarios. are we going to cover them somehow?

File: Kentico.Kontent.Delivery.Caching/Extensions/ServiceCollectionExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        }
+
+        private static IServiceCollection RegisterDependencis(this IServiceCollection services)
+        {
+            services.TryAddSingleton<IDeliveryCacheManager, DeliveryCacheManager>();
+            services.TryAddSingleton<IMemoryCache, MemoryCache>();
+
+            return services;
+        }
+
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery.Caching/Extensions/ServiceCollectionExtensions.cs]: empty lines

File: Kentico.Kontent.Delivery.Caching/Extensions/ServiceCollectionExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+                    return new ConfigureNamedOptions<DeliveryClientFactoryOptions>(name, o =>
+                    {
+                        var client = o.DeliveryClientsActions.FirstOrDefault()?.Invoke();
+                        o.DeliveryClientsActions.Add(() =>
+                        {
+                            var deliveryCacheManager = sp.GetRequiredService<IDeliveryCacheManager>();
+                            return new DeliveryClientCache(deliveryCacheManager, client);
+                        });
+                    });
+
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery.Caching/Extensions/ServiceCollectionExtensions.cs]: lines

File: Kentico.Kontent.Delivery.Caching.Tests/DeliveryCacheManager.cs [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+                   var result = memoryCache.TryGetValue(c.Arguments[0], out var value);
+                   return result;
+
+               })
+               .AssignsOutAndRefParametersLazily(c => {
+                   var result = memoryCache.TryGetValue(c.Arguments[0], out var value);
+                   return new object[] { value };
+               });
+
+
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery.Caching.Tests/DeliveryCacheManager.cs]: lines

File: Kentico.Kontent.Delivery.Caching.Tests/DeliveryCacheManager.cs [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Kentico.Kontent.Delivery.Caching.Tests
+{
+    public class DeliveryCacheManagerTests
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery.Caching.Tests/DeliveryCacheManager.cs]: please rename the file accordingly (DeliveryCacheManagerTests.cs)

File: Kentico.Kontent.Delivery.Caching/DeliveryCacheManager.cs [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+                // Set different timeout for stale content
+                var valueCacheOptions = new MemoryCacheEntryOptions();
+                if (value is AbstractResponse ar && ar.HasStaleContent)
+                {
+                    valueCacheOptions.SetAbsoluteExpiration(_cacheOptions.StaleContentExpiration);
+                }
+                else
+                {
+                    valueCacheOptions.SetSlidingExpiration(_cacheOptions.DefaultExpiration);
+                }
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery.Caching/DeliveryCacheManager.cs]: the original implementation of the time-bound cache manager always set absolute expiration. but I think this is fine.

**Comment by petrsvihlik** [comment]
once all issues are resolved, we'll release a beta and give it a try in https://github.com/Kentico/kontent-boilerplate-net/

File: Kentico.Kontent.Delivery.Tests/Extensions/FakeServiceCollection.cs (hunk 1 above)
Comment (tomasjurasek) [reply Kentico.Kontent.Delivery.Tests/Extensions/FakeServiceCollection.cs]: With FaceServiceCollection implementation is more probably we can do a mistake with this "faking" implementation. We need more realistic tests with ServiceCollection. I did some basic scenarios with it. In the future, we can do more scenarios

Comment (petrsvihlik) [reply Kentico.Kontent.Delivery.Tests/Extensions/FakeServiceCollection.cs]: ok

File: Kentico.Kontent.Delivery.Caching/DeliveryCacheManager.cs [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+                        }
+                    }
+
+                    if (tokenSource != null)
+                    {
+                        valueCacheOptions.AddExpirationToken(new CancellationChangeToken(tokenSource.Token));
+                    }
+                }
+
+                return _memoryCache.Set(key, value, valueCacheOptions);
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery.Caching/DeliveryCacheManager.cs]: @tomasjurasek 

Just out of curiosity - wouldn't it be better to use `_memoryCache.GetOrCreateAsync()`? 

- https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.caching.memory.cacheextensions.getorcreateasync?view=dotnet-plat-ext-3.1
- https://docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-3.1
- https://www.hanselman.com/blog/EyesWideOpenCorrectCachingIsAlwaysHard.aspx

**Pull request merged**